Add support for browser.tabs script APIs.

kiaraarose opened the pull request:
#### ea26ecd46cd6578b6dc5d66fddce4ac9664108c3
<pre>
Add support for browser.tabs script APIs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=262816">https://bugs.webkit.org/show_bug.cgi?id=262816</a>

Reviewed by Timothy Hatcher.

This patch implements the tabs.executeScript(), tabs.insertCSS(), and tabs.removeCSS() APIs.

Testing:
- Calls to methods with an invalid tab results in an error.
- A result is returned from for each frame for scripting.executeScript().
- Injections execute in the top frame only if allFrames isn&apos;t specified or is false.

Will add more extensive tests in a follow up PR.

* Source/WebKit/Shared/Extensions/WebExtensionDynamicScripts.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionParameters.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsExecuteScript):
(WebKit::WebExtensionContext::tabsInsertCSS):
(WebKit::WebExtensionContext::tabsRemoveCSS):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::toInjectionResultParameters):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h:
(WebKit::WebExtensionDynamicScripts::InjectionResultHolder::create):
(WebKit::WebExtensionDynamicScripts::InjectionResultHolder::InjectionResultHolder):
Created a ref counted class to hold the results of the script injections so that the data is still
valid when the callback aggregator is called.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::toWebAPI):
(WebKit::WebExtensionAPITabs::parseScriptOptions):
(WebKit::WebExtensionAPITabs::executeScript):
(WebKit::WebExtensionAPITabs::insertCSS):
(WebKit::WebExtensionAPITabs::removeCSS):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITabs.idl:

Canonical link: <a href="https://commits.webkit.org/269028@main">https://commits.webkit.org/269028@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cadd6b1ba35877a0ffecba1f8a2f61cb82bae05b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21377 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/21689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22429 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23238 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/19816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24986 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21932 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21601 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21251 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18521 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24090 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18419 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19499 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19577 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23541 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/20066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19382 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/23628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2651 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19970 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -136,6 +136,22 @@ void removeStyleSheets(SourcePairs styleSheetPairs, WebCore::UserContentInjected
+        parameters.result = resultOfExecution;
+
+    if (info)
+        parameters.frameID = info.isMainFrame ? WebExtensionFrameConstants::MainFrameIdentifier : toWebExtensionFrameIdentifier(info._handle.frameID);
```
`parameters.frameID = toWebExtensionFrameIdentifier(info)`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
```diff
@@ -961,6 +1026,71 @@ bool isValid(std::optional<WebExtensionTabIdentifier> identifier, NSString **out
+        }
+        callback->call(toWebAPI(results.value()));
+    }, extensionContext().identifier());
+
```
Remove empty line.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
```diff
@@ -961,6 +1026,71 @@ bool isValid(std::optional<WebExtensionTabIdentifier> identifier, NSString **out
+            callback->reportError(error.value());
+            return;
+        }
+        callback->call(toWebAPI(results.value()));
```
Add newline above.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
```diff
@@ -961,6 +1026,71 @@ bool isValid(std::optional<WebExtensionTabIdentifier> identifier, NSString **out
+            callback->reportError(error.value());
+            return;
+        }
+        callback->call();
```
Add newline above.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
```diff
@@ -961,6 +1026,71 @@ bool isValid(std::optional<WebExtensionTabIdentifier> identifier, NSString **out
+        }
+        callback->call();
+    }, extensionContext().identifier());
+
```
Remove empty line.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
```diff
@@ -961,6 +1026,71 @@ bool isValid(std::optional<WebExtensionTabIdentifier> identifier, NSString **out
+            callback->reportError(error.value());
+            return;
+        }
+        callback->call();
```
Add newline above.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
```diff
@@ -542,6 +551,119 @@ static inline String toMIMEType(WebExtensionTab::ImageFormat format)
+
+    std::optional<SourcePair> scriptData;
+    if (parameters.code)
+        scriptData = make_pair(parameters.code.value(), std::nullopt);
```
`scriptData = SourcePair { parameters.code.value(), std::nullopt };`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
```diff
@@ -542,6 +551,119 @@ static inline String toMIMEType(WebExtensionTab::ImageFormat format)
+        completionHandler(injectionResults->results, std::nullopt);
+    });
+
+    [webView _frames:makeBlockPtr([this, protectedThis = Ref { *this }, webView = RetainPtr { webView }, tab = RefPtr { tab }, injectionResults, aggregator = WTFMove(aggregator), parameters = WTFMove(parameters)](_WKFrameTreeNode *mainFrame) mutable {
```
Should be able to do `tab` instead of `tab = RefPtr { tab }` and `aggregator` instead of `aggregator = WTFMove(aggregator)`. Ditto for `parameters` instead of `parameters = WTFMove(parameters)`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
```diff
@@ -542,6 +551,119 @@ static inline String toMIMEType(WebExtensionTab::ImageFormat format)
+            }
+
+            for (auto& script : scriptPairs) {
+                [webView _evaluateJavaScript:script.value().first withSourceURL:script.value().second.value_or(URL { }) inFrame:info inContentWorld:m_contentScriptWorld.get()->wrapper() completionHandler:makeBlockPtr([injectionResults, aggregator] (id resultOfExecution, NSError *error) mutable {
```
Remove space between `] (`.

kiaraarose committed:
Add support for browser.tabs script APIs.
https://bugs.webkit.org/show_bug.cgi?id=262816

Reviewed by Timothy Hatcher.

This patch implements the tabs.executeScript(), tabs.insertCSS(), and tabs.removeCSS() APIs.

Testing:
- Calls to methods with an invalid tab results in an error.
- A result is returned from for each frame for scripting.executeScript().
- Injections execute in the top frame only if allFrames isn't specified or is false.

Will add more extensive tests in a follow up PR.

* Source/WebKit/Shared/Extensions/WebExtensionDynamicScripts.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionScriptInjectionParameters.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsExecuteScript):
(WebKit::WebExtensionContext::tabsInsertCSS):
(WebKit::WebExtensionContext::tabsRemoveCSS):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::toInjectionResultParameters):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionDynamicScripts.h:
(WebKit::WebExtensionDynamicScripts::InjectionResultHolder::create):
(WebKit::WebExtensionDynamicScripts::InjectionResultHolder::InjectionResultHolder):
Created a ref counted class to hold the results of the script injections so that the data is still
valid when the callback aggregator is called.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::toWebAPI):
(WebKit::WebExtensionAPITabs::parseScriptOptions):
(WebKit::WebExtensionAPITabs::executeScript):
(WebKit::WebExtensionAPITabs::insertCSS):
(WebKit::WebExtensionAPITabs::removeCSS):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPITabs.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPITabs.idl:

Canonical link: https://commits.webkit.org/269028@main
